File: src/main/java/at/searles/regexp/Text.java
package at.searles.regexp;


class Text implements Regexp {

    private final String string;

    Text(CharSequence seq) {
        this.string = seq.toString();
    }

    @Override
    public String toString() {
        return string;
    }

    @Override
    public <A> A accept(Visitor<A> visitor) {
        return visitor.visitText(string);
    }
}

File: src/main/java/at/searles/regexp/Text.kt
package at.searles.regexp

class Text(seq: CharSequence) : Regexp {
    private val string: String = seq.toString()

    override fun toString(): String {
        return string
    }

    override fun <A> accept(visitor: Visitor<A>): A {
        return visitor.visitText(string)
    }

    companion object {
        fun many(vararg seqs: CharSequence): Regexp {
            return seqs.drop(1).fold(Text(seqs.first()) as Regexp, { a, b -> a or Text(b) } )
        }
    }
}